[Federation] Refactor sync controller's reconcile method for maintainability

marun opened the pull request:
This PR refactors the sync controllers reconcile method for maintainability with the goal of eliminating the need for type-specific controller unit tests.  The unit test coverage for reconcile is not complete, but I think it's a good start.

cc: @kubernetes/sig-federation-pr-reviews

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45374)
<!-- Reviewable:end -->

marun commented:
@shashidharatd Returning status from reconcile follows from your recent changes to the services controller.  Reviewing that work again, I should probably centralize the status codes for reuse rather than defining yet another set.

nikhiljindal commented:
@perotinus already has a PR that does the same change of reconcile() returning status codes and moving out delivery logic out from that method.
@emaildanwilson also has a PR already that does the same change of removing the logic of what operations to perform into a separate method to make that easier to test.

marun committed:
fed: Further refactor of sync controller reconcile

This change breaks out non-delete cluster updates into a method for
testability.

marun commented:
@nikhiljindal Yeah, looks like we're all colliding.  My bad for not keeping up with review.  

@perotinus @emaildanwilson would it makes sense to arrange a time to chat about resolving these conflicts?

Regardless of whether this merges, I'd like to get some feedback from others as to the viability of the proposed approach to testing.  My goal is to reduce the scope of reconcile to chunks that can be tested in isolation.  #45382 doesn't include test changes, and I think the tests proposed in #40234 are too heavyweight.  That said, the approach I've taken here is bordering on functional to allow the exercise of every possible code path.  While I like the fact that this enables 100% coverage, I'm not entirely happy with the impact on readability.

perotinus commented:
@marun @emaildanwilson I'm happy to sit down and work through this. We are all touching on the same thing in a slightly different way. Do you want to set up some time today or tomorrow to chat?

emaildanwilson commented:
@perotinus yes. happy to chat as well to get this resolved. I can make myself available just about any time.

marun commented:
rebased

@k8s-bot pull-kubernetes-federation-e2e-gce test this
status: http://prow.k8s.io/?type=presubmit&pull=45374&job=pull-kubernetes-federation-e2e-gce

marun commented:
rebased

@k8s-bot pull-kubernetes-federation-e2e-gce test this
status: http://prow.k8s.io/?type=presubmit&pull=45374&job=pull-kubernetes-federation-e2e-gce

marun commented:
presubmit is passing

perotinus commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -233,6 +227,14 @@ func (s *FederationSyncController) Run(stopChan <-chan struct{}) {
 	go wait.Until(s.worker, time.Second, stopChan)
 
 	util.StartBackoffGC(s.backoff, stopChan)
+	// Ensure all goroutines are cleaned up when the stop channel closes
```
Nit: blank line above this one?

marun replied:
Done

perotinus commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -320,126 +322,158 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
-	glog.V(3).Infof("Ensuring delete object from underlying clusters finalizer for %s: %s",
-		kind, namespacedName)
-	// Add the required finalizers before creating the resource in underlying clusters.
+	glog.V(3).Infof("Ensuring finalizers for %s %q", kind, key)
```
Nit: "Ensuring finalizers exist on %s %q

Or something else? This wording of "Ensure finalizers for" has always confused me a little bit.

marun replied:
I'm not sure I understand your concern.  How is ``Ensuring finalizers on secret "myns/foo"`` semantically different from ``Ensuring finalizers for secret "myns/foo"``?

perotinus replied:
"exist" was the key addition: if I didn't know what a finalizer was, I might assume that it was something that was created somewhere else and triggered when the object was being destroyed. Saying "finalizers exist on" is more like what's actually happening, and makes it easier for someone not familiar with finalizers to understand that they're being added to the object, IMO.

marun replied:
Fair enough, I'll update it.  I was assuming that a given maintainer would know what finalizers were.

perotinus commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -320,126 +322,158 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
-	glog.V(3).Infof("Syncing %s %s in underlying clusters", kind, namespacedName)
+	return syncToClusters(
+		s.informer.GetReadyClusters,
+		func(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object) ([]util.FederatedOperation, error) {
```
Consider making this a local variable? It might be better in a text editor, but the wrapping on GitHub's diff view makes this a bear to read.

marun replied:
Done

marun committed:
fed: Factor annotation mutation out of the crudtester for reuse

marun committed:
fed: Refactor sync controller's reconcile method

marun commented on federation/pkg/federatedtypes/adapter.go:
```diff
@@ -61,3 +61,12 @@ type FederatedTypeAdapter interface {
 type AdapterFactory func(client federationclientset.Interface) FederatedTypeAdapter
+
+// SetAnnotation sets the given key and value in the given object's ObjectMeta.Annotations map
+func SetAnnotation(adapter FederatedTypeAdapter, obj pkgruntime.Object, key, value string) {
```
Done

marun commented:
addressed comments

@k8s-bot pull-kubernetes-federation-e2e-gce test this
status: http://prow.k8s.io/?type=presubmit&pull=45374&job=pull-kubernetes-federation-e2e-gce

madhusudancs commented:
@nikhiljindal reassigning this to you since you have already looked at this PR.

/assign @nikhiljindal 
/unassign

Please ping me on IM or something if you want me to take a look.

perotinus commented:
/lgtm

marun commented:
addressed comment

@k8s-bot pull-kubernetes-federation-e2e-gce test this
status: http://prow.k8s.io/?type=presubmit&pull=45374&job=pull-kubernetes-federation-e2e-gce

marun commented:
@nikhiljindal Review appreciated to unblock #40234 for @emaildanwilson

nikhiljindal commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -320,113 +323,90 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
+	// Create a copy before modifying the resource to prevent racing with other readers.
+	copiedObj, err := api.Scheme.DeepCopy(cachedObj)
+	if err != nil {
+		return nil, fmt.Errorf("Error in retrieving %s %q from store: %v", kind, key, err)
```
We always start the error with smaller case since the calling function can prepend more context.

marun replied:
That's not the case here, though, the only caller is just glog'ing the error directly.

nikhiljindal commented on federation/pkg/federation-controller/sync/controller.go:
```diff
@@ -320,113 +323,90 @@ func (s *FederationSyncController) reconcile(namespacedName types.NamespacedName
+		return nil, fmt.Errorf("Error in retrieving %s %q from store: %v", kind, key, err)
+	}
+	if !s.adapter.IsExpectedType(copiedObj) {
+		return nil, fmt.Errorf("Object is not the expected type: %v", copiedObj)
```
same. start with lower case

marun replied:
same

nikhiljindal commented:
I will let @perotinus and @emaildanwilson review this one in detail.
/approve

emaildanwilson commented:
I reviewed this. looks great!

marun committed:
fed: Update sync controller to use runtime.HandleError

marun commented:
@nikhiljindal I've added a new commit that switches to using ``runtime.HandleError`` at the site of the error rather than bubbling up and using ``glog.Errorf``.  I'm hoping this makes it clear that the error messages aren't intended to be consumed outside of the returning functions.

perotinus commented:
Still /lgtm

marun commented:
updated bazel

marun commented:
@nikhiljindal PTAL to unblock @emaildanwilson and @perotinus's follow-on PRs.   And feel free to require the case change you asked for instead of my alternate.

perotinus commented:
/lgtm
